L1Trigger/L1THGCalUtilities/plugins/CaloTruthCellsProducer.cc

```diff
@@ -0,0 +1,397 @@
+          edm::LogError("CaloTruthCellsProducer") << ex.what();
+          continue;
+        }
+
+        // tcMap[tcId] += hitMap[hitId.rawId()] * hAndF.first;
+        tcToCalo.emplace(tcId, ref);
+      }
+    }
+
+    // ordered by the gen particle index
+    int genIndex(caloParticle.g4Tracks().at(0).genpartIndex() - 1);
+    if (genIndex < 0) // shouldn't happen
+      continue;
+    if (genIndex >= int(orderedCaloRefs.size()))
+      orderedCaloRefs.resize(genIndex + 1);
+
+    orderedCaloRefs[genIndex] = ref;
+  }
+
+  auto outMap(std::make_unique<CaloToCellsMap>(caloParticlesHandle, triggerCellsHandle));
+  std::unique_ptr<l1t::HGCalTriggerCellBxCollection> outCollection;
+  if (makeCellsCollection_)
+    outCollection.reset(new l1t::HGCalTriggerCellBxCollection);
+
+  int bx(triggerCells.getFirstBX());
+  unsigned bxOffset(0);
+
+  typedef edm::Ptr<l1t::HGCalTriggerCell> TriggerCellPtr;
+  typedef edm::Ptr<l1t::HGCalCluster> ClusterPtr;
+
+  // ClusteringDummyImpl only considers BX 0, so we dump all cells to one vector
+  std::vector<TriggerCellPtr> triggerCellPtrs;
+
+  // loop through all bunch crossings
+  for (unsigned iC(0); iC != triggerCells.size(); ++iC) {
+    if (iC >= bxOffset + triggerCells.size(bx)) {
+      bxOffset += triggerCells.size(bx);
+      bx += 1;
+    }
+
+    auto& cell(triggerCells[iC]);
+
+    auto mapElem(tcToCalo.find(cell.detId()));
+    if (mapElem == tcToCalo.end())
+      continue;
+
+    edm::Ref<l1t::HGCalTriggerCellBxCollection> ref(triggerCellsHandle, iC);
+
+    outMap->insert(mapElem->second, ref);
+
+    if (makeCellsCollection_)
+      outCollection->push_back(bx, cell);
+
+    triggerCellPtrs.emplace_back(triggerCellsHandle, iC);
+  }
+
+  _event.put(std::move(outMap));
+  if (makeCellsCollection_)
+    _event.put(std::move(outCollection), "HGCalVFEProcessorSums");
+
+  auto outClusters(std::make_unique<l1t::HGCalClusterBxCollection>());
+
+  auto sortCellPtrs([](TriggerCellPtr const& lhs, TriggerCellPtr const& rhs)->bool {
+      return lhs->mipPt() > rhs->mipPt();
+    });
+
+  std::sort(triggerCellPtrs.begin(), triggerCellPtrs.end(), sortCellPtrs);
+  dummyClustering_.clusterizeDummy(triggerCellPtrs, *outClusters);
+
+  std::map<CaloParticleRef, std::vector<unsigned>> caloToClusterIndices;
+  for (unsigned iC(0); iC != outClusters->size(); ++iC) {
+    auto& cluster((*outClusters)[iC]);
+    // cluster detId and cell detId are identical
+    auto caloRef(tcToCalo.at(cluster.detId()));
+    caloToClusterIndices[caloRef].push_back(iC);
+  }
+
+  auto&& clustersHandle(_event.put(std::move(outClusters), "HGCalBackendLayer1Processor2DClustering"));
+
+  auto outMulticlusters(std::make_unique<l1t::HGCalMulticlusterBxCollection>());
+
+  for (auto& ref : orderedCaloRefs) {
+    if (ref.isNull()) // shouldn't happen
+      continue;
+
+    auto& caloParticle(*ref);
+
+    l1t::HGCalTriggerCell dummyCell;
+    dummyCell.setDetId(caloParticle.g4Tracks().at(0).genpartIndex() - 1);
```

Comment:
Instead of adding these dummy TCs and clusters, could add a `setDetId` method to the cluster class.

Reply:
OK, with your permission :)